File: frontend/src/services/rawMaterialService.ts
import { api } from "@/services/api"
import type { PageResponse } from "@/services/types"

export type RawMaterial = {
  id: string
  name: string
  description?: string | null
  price: number
  stockQuantity: number
  unit: string
  createdAt: string
  updatedAt: string
}

export type RawMaterialRequest = {
  name: string
  description?: string
  price: number
  stockQuantity: number
  unit: string
}

export const rawMaterialService = {
  async list(params: { page: number; size: number; search?: string | null }) {
    const res = await api.get<PageResponse<RawMaterial>>("/raw-materials", {
      params: {
        page: params.page,
        size: params.size,
        search: params.search || undefined,
      },
    })
    return res.data
  },
  async listAll() {
    const res = await api.get<RawMaterial[]>("/raw-materials/all")
    return res.data
  },
  async get(id: string) {
    const res = await api.get<RawMaterial>(`/raw-materials/${id}`)
    return res.data
  },
  async create(body: RawMaterialRequest) {
    const res = await api.post<RawMaterial>("/raw-materials", body)
    return res.data
  },
  async update(id: string, body: RawMaterialRequest) {
    const res = await api.put<RawMaterial>(`/raw-materials/${id}`, body)
    return res.data
  },
  async remove(id: string) {
    await api.delete(`/raw-materials/${id}`)
  },
}

File: backend/src/main/java/com/duckstock/service/RawMaterialService.java
package com.duckstock.service;

import com.duckstock.dto.common.PageResponse;
import com.duckstock.dto.rawmaterial.RawMaterialRequest;
import com.duckstock.dto.rawmaterial.RawMaterialResponse;
import com.duckstock.entity.RawMaterial;
import com.duckstock.exception.BusinessException;
import com.duckstock.exception.ResourceNotFoundException;
import io.quarkus.hibernate.orm.panache.PanacheQuery;
import io.quarkus.panache.common.Page;
import io.quarkus.panache.common.Sort;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@ApplicationScoped
public class RawMaterialService {

    public PageResponse<RawMaterialResponse> listAll(int page, int size, String search) {
        PanacheQuery<RawMaterial> query;

        if (search != null && !search.isBlank()) {
            query = RawMaterial.find("LOWER(name) LIKE LOWER(?1)", Sort.descending("createdAt"),
                    "%" + search.trim() + "%");
        } else {
            query = RawMaterial.findAll(Sort.descending("createdAt"));
        }

        long totalElements = query.count();
        List<RawMaterialResponse> content = query.page(Page.of(page, size))
                .list()
                .stream()
                .map(RawMaterialResponse::from)
                .collect(Collectors.toList());

        return new PageResponse<>(content, page, size, totalElements);
    }

    public RawMaterialResponse findById(UUID id) {
        RawMaterial rawMaterial = RawMaterial.findById(id);
        if (rawMaterial == null) {
            throw new ResourceNotFoundException("Raw material not found with id: " + id);
        }
        return RawMaterialResponse.from(rawMaterial);
    }

    @Transactional
    public RawMaterialResponse create(RawMaterialRequest request) {
        RawMaterial rawMaterial = new RawMaterial();
        rawMaterial.name = request.name;
        rawMaterial.description = request.description;
        rawMaterial.price = request.price;
        rawMaterial.stockQuantity = request.stockQuantity;
        rawMaterial.unit = request.unit;
        rawMaterial.persist();
        return RawMaterialResponse.from(rawMaterial);
    }

    @Transactional
    public RawMaterialResponse update(UUID id, RawMaterialRequest request) {
        RawMaterial rawMaterial = RawMaterial.findById(id);
        if (rawMaterial == null) {
            throw new ResourceNotFoundException("Raw material not found with id: " + id);
        }
        
        if (request.name != null) {
            rawMaterial.name = request.name;
        }
        if (request.description != null) {
            rawMaterial.description = request.description;
        }
        if (request.price != null) {
            rawMaterial.price = request.price;
        }
        if (request.stockQuantity != null) {
            rawMaterial.stockQuantity = request.stockQuantity;
        }
        if (request.unit != null) {
            rawMaterial.unit = request.unit;
        }
        
        rawMaterial.persist();
        return RawMaterialResponse.from(rawMaterial);
    }

    @Transactional
    public void delete(UUID id) {
        RawMaterial rawMaterial = RawMaterial.findById(id);
        if (rawMaterial == null) {
            throw new ResourceNotFoundException("Raw material not found with id: " + id);
        }
        rawMaterial.delete();
    }

    public List<RawMaterialResponse> listAllNoPagination() {
        return RawMaterial.findAll(Sort.ascending("name"))
                .list()
                .stream()
                .map(rm -> RawMaterialResponse.from((RawMaterial) rm))
                .collect(Collectors.toList());
    }
}
